lightning/src/ln/channelmanager.rs

```diff
@@ -2059,8 +2111,37 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		} else if has_err {
 			Err(PaymentSendFailure::AllFailedRetrySafe(results.drain(..).map(|r| r.unwrap_err()).collect()))
 		} else {
-			Ok(())
+			Ok(payment_id)
+		}
+	}
+
+	/// XXX
+	pub fn retry_payment(&self, route: &Route, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>, payment_id: PaymentId) -> Result<(), PaymentSendFailure> {
+		let outbounds = self.pending_outbound_payments.lock().unwrap();
+		let mut total_msat_opt = None;
+		if let Some(payment) = outbounds.get(&payment_id) {
+			match payment {
+				PendingOutboundPayment::Retryable { total_msat, payment_hash: hash, .. } => {
+					// XXX checks that total_msat_opt makes sense
+					total_msat_opt = Some(*total_msat);
+					if payment_hash != *hash {
+						return Err(PaymentSendFailure::ParameterError(APIError::APIMisuseError {
+							err: "Payment hash does not match given payment ID".to_string()
+						}))
+					}
+				},
+				PendingOutboundPayment::Legacy { .. } => {
+					return Err(PaymentSendFailure::ParameterError(APIError::APIMisuseError {
+						err: "Unable to retry payments that were serialized as legacy payments".to_string()
+					}))
+				}
+			}
+		} else {
+			// XXX they could retry with a bogus amt/hash/secret/payment_id here and we wouldn't catch it atm
+			// should we keep payments in pending_outbound_payments until they succeed/expire/manually removed?
```

Comment:
Hmm, I was kinda figuring we'd require users just call `send_payment` again, though I admit it would be kinda slick to let them just call the same method here. I guess we could do something like time out pending all-paths-failed payments after, eg, 3 blocks or so? Do you have an opinion @jkczyz?

Reply:
Yea, I was thinking we'd have separate interfaces, but maybe the memory usage is not so bad? Like, we'd have a bunch of memory used for failed payments, but we could time-box it. For high-volume scenarios it could be a real concern, but we'd clean up any memory (a) after the payment succeeds, even across a different path, or (b) after some time. Maybe we can figure out how to address the DoS later?

Reply:
We could parameterize `send_payment` further by the invoice expiry, and remove pending outbounds at `min(invoice_expiry, 3 blocks)` maybe?

In any case, given this week's a bit busy, I'm in favor of punting complete payment retries to follow-up

Reply:
> In any case, given this week's a bit busy, I'm in favor of punting complete payment retries to follow-up

Well we either have to handle it here or in #1059, both for 0.0.102. I suppose we *can* punt "tightly limit expiry time" to a later release, even if we do expire here?

Reply:
Oh, I meant punt complete payment retries via the `retry_payment` API. I think #1059 already does complete payment retries, just via `send_payment`

Reply:
> Oh, I meant punt complete payment retries via the retry_payment API. I think #1059 already does complete payment retries, just via send_payment


Right, my point was its probably ~the same amount of code to handle switching between `retry_payment` and `send_payment` in #1059 as it is to just call `retry_payment` instead of `send_payment` and not remove the payment entry until a payment succeeds in this PR :)